File: 2018/07-jul/7-29-overloading/php/Garden.php
<?php
/**
 * Representation of a Garden I do work in at Tod's Point
 * User: Andrew Jarombek
 * Date: 7/28/2018
 */

class Garden
{

    private $animals;
    private $plants;
    private $randomObjects;

    /**
     * Garden constructor that takes in three lists of items in the garden.
     * These lists default to an empty list if no argument is passed to the constructor.
     * @param array $animals - a list of animals in the garden
     * @param array $plants - a list of plants in the garden
     * @param array $randomObjects - a list of random objects in the garden
     */
    public function __construct($animals=[], $plants=[], $randomObjects=[])
    {
        $this->animals = $animals;
        $this->plants = $plants;
        $this->randomObjects = $randomObjects;
    }

    /**
     * Checks whether a certain item exists in the garden.  Depending on the type of the argument
     * a different internal array will be searched for the item.
     * @param $item - an Animal, Plant, or any other random object
     * @return bool - true if the item exists, false otherwise
     */
    public function inGarden($item) {
        if ($item != null) {

            if ($item instanceof Animal) {
                return in_array($item, $this->animals);
            } else if ($item instanceof Plant) {
                return in_array($item, $this->plants);
            } else {
                return in_array($item, $this->randomObjects);
            }
        } else {
            return false;
        }
    }
}

File: 2018/07-jul/7-29-overloading/python/Garden.py
"""
Representation of a Garden I do work in at Tod's Point
Author: Andrew Jarombek
Date: 7/28/2018
"""

from Animal import Animal
from Plant import Plant


class Garden(object):

    def __init__(self, animals=list(), plants=list(), random_objects=list()):
        """
        Constructor for a garden which contains animals, plants, and some other random objects
        :param animals: a list of animals in the garden (defaults to an empty list)
        :param plants: a list of plants in the garden (defaults to an empty list)
        :param random_objects: a list of random objects in the garden (defaults to an empty list)
        """
        self.animals = animals
        self.plants = plants
        self.random_objects = random_objects

    def in_garden(self, item):
        """
        Check whether an item exists in the garden.  The item can be a plant, animal, or any
        random object.  Depending on the type of the object, a different internal list will be
        checked for the existence of the object.
        :param item: the item to check for existence in the garden
        :return: False if the item is None or doesnt exist in the garden, True otherwise
        """

        if item is not None:
            if type(item) is Animal:
                return item in self.animals
            if type(item) is Plant:
                return item in self.plants
            else:
                return item in self.random_objects
        else:
            return False

    def __str__(self):
        """
        Create and return a human readable string that represents the Garden object
        :return: string
        """
        return '({}, {}, {})'.format(self.animals, self.plants, self.random_objects)

    def __repr__(self):
        """
        Create an internal representation of the Garden object.
        Simply delegate to the __str__() special method
        :return: string
        """
        return str(self)
